File: src/main/java/leetcode_30_days_challenge/day4/LC283MoveZeroes.java
package leetcode_30_days_challenge.day4;

public class LC283MoveZeroes {
    //    Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.
//
//    Example:
//
//    Input: [0,1,0,3,12]
//    Output: [1,3,12,0,0]
    public void moveZeroes(int[] nums) {
        int i = 0;
        int j = 0;

        while (j < nums.length) {
            if (nums[j] != 0) {
                nums[i++] = nums[j];
            }
            j++;
        }

        while (i < nums.length) {
            nums[i++] = 0;
        }
    }
}

File: src/main/scala/leetcode_challenges/april/day4/LC283_MoveZeroes.scala
package leetcode_challenges.april.day4

object LC283_MoveZeroes {
  //  Given an array nums, write a function to move all 0's to the end of it while maintaining
  // the relative order of the non-zero elements.
  //
  //  Example:
  //
  //    Input: [0,1,0,3,12]
  //  Output: [1,3,12,0,0]
  def moveZeroes(nums: Array[Int]): Unit = {
    var j = 0
    for (i <- nums.indices) {
      j = nums.indexWhere(_ != 0, j)
      if (j < 0) {
        for (k <- i until nums.length) nums(k) = 0
        return
      }
      nums(i) = nums(j)
      j += 1
    }
  }
}
